http2, async-wrap: introduce AliasedBuffer class

mike-kaufman opened the pull request:
This change introduces an `AliasedBuffer` class and updates async-wrap
and http2 to use this class.

A common technique to optimize performance is to create a native buffer
and then map that native buffer to user space via JS array.  The runtime
can efficiently write to the native buffer without having to route
though JS, and the values being written are accessible from user space.

While efficient, this technique allows modifications to user
space memory w/out going through JS type system APIs, effectively
bypassing any monitoring the JS VM has in place to track program state
modifications.  The result is that monitors have an incorrect view
of program state.

The AliasedBuffer class provides a future placeholder where this
technique can be used, but writes can still be observed.  To achieve
this, the node-chakra-core fork will add in appropriate tracking logic
in the `AliasedBuffer`'s `SetValue()` method. Going forward, this class can
evolve to support more sophisticated mechanisms if necessary.

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included  (*tests added, happy to run benchmarks, just lmk*)
- [ ] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
async-wrap, http2

mike-kaufman commented:
/cc @jasnell, @addaleax, @mrkmarron.

mike-kaufman commented on src/aliased_buffer.h:
```diff
@@ -0,0 +1,193 @@
+   */
+  inline void SetValue(const int index, NativeT value) {
+    // if (index >= count_) {
+    // TODO(@mike-kaufman): Assert?  Fail fast?
```
Wasn't clear if I should do anything here.  Please let me know if there's existing precedent around such checks.

jasnell replied:
`CHECK` is likely appropriate. /cc @bnoordhuis

mike-kaufman replied:
added a CHECK.

mike-kaufman commented on src/aliased_buffer.h:
```diff
@@ -0,0 +1,193 @@
+   */
+  inline const NativeT GetValue(const int index) const {
+    // if (index >= count_) {
+    // TODO(@mike-kaufman): Assert?  Fail fast?
```
similiar here...

mike-kaufman replied:
fixed now with a CHECK.

mike-kaufman commented on src/aliased_buffer.h:
```diff
@@ -0,0 +1,193 @@
+      free(this->buffer_);
+    }
+    // TODO(@mike-kaufman):  what do we do about freeing the
+    // v8::Global<ArrayBuffer> & the v8::Global<U>
```
not clear to me if I need to do anything to manage lifecycle of a `v8::Global`?  Do I need to call `reset()`?

jasnell replied:
That I'm unsure about. @addaleax or @bnoordhuis may know.

trevnorris replied:
A `Global<>` does need to be reset.

mike-kaufman replied:
fixed. calling `.Reset()` on the v8::Global.

mike-kaufman commented on src/node_http2_state.h:
```diff
@@ -0,0 +1,116 @@
+#ifndef SRC_NODE_HTTP2_STATE_H_
```
I had to move most of the stuff here out of `node_http2.cc` because `http2_state` was forward declaration, and I was seeing a compiler warning about calling delete on a forward declaration.  I initially tried to introduce a `#include "node_http2.h"` into `env.h`, but that was causing cycles.

mike-kaufman commented on test/cctest/test_aliased_buffer.cc:
```diff
@@ -0,0 +1,205 @@
+    // v8::MaybeLocal<v8::Number> v3 = v2->ToNumber();
+    // v8::Local<v8::Number> v4 = v3.ToLocalChecked();
+    // NativeT actualValue = *(reinterpret_cast<NativeT*>(&v4->Value()));
+    // EXPECT_TRUE((NativeT)(v4->Value()) == oracle[i]);
```
I want to validate the value being read from the JS api is the same as the native type value, but I couldn't get this to work.   Is there an easy way to do this?  Some casting magic I need to do?

jasnell replied:
Not that I'm aware of.

jasnell commented on src/aliased_buffer.h:
```diff
@@ -0,0 +1,193 @@
+
+    // allocate native buffer
+    this->buffer_ = reinterpret_cast<NativeT*>(
+      calloc(count, sizeof(NativeT)));
```
Could this follow the same pattern as `MaybeStackBuffer`... or even make use of that to avoid the `calloc` where possible?

mike-kaufman replied:
I could, but I'm not following why.  Can you help me understand your concerns w/ using calloc?  I was trying to be consistent w/ original buffers that were alloc'd on heap.

jasnell replied:
Consistency would be the main reason but it's certainly not critical. I don't think the performance difference is going to be enough to worry about.

mike-kaufman replied:
I'm now using `node::UncheckedCalloc`.  Is using `MaybeStackBuffer` still something to investigate?

jasnell commented:
Very good. I'll give it a thorough review and sign off later but I'd like to get @addaleax, @bnoordhuis and @trevnorris to take a look also

mscdex commented:
What kind of performance impact does this have?

mike-kaufman commented:
> What kind of performance impact does this have?

@mscdex - what tests would you suggest I run to measure this?

I ran acmeair-nodejs benchmark, comparing my local master (at [this commit](https://github.com/nodejs/node/commit/689a64319864433b32235b9d6ac4889f4cdcfea5)) with my changes.  

Results didn't show any meaningful impact:

master:
```
    "numRequests": 533758,
    "elapsedTime": 209787,
    "requestsPerSecond": 2544.2853942331985,
    "averageLatency": 9.682236144469966,
```

with my change:
```
    "numRequests": 536971,
    "elapsedTime": 209800,
    "requestsPerSecond": 2559.4423260247854,
    "averageLatency": 9.625374554677999,
  ```

Note that intention with this change was to have the `GetValue()/SetValue()` methods inline and effectively result in 0/minimal overhead vs previous code (i.e., directly accessing offsets in the native buffer).  Now, the use the overloaded operator [] may add some overhead here (it is a bit more complex). 

I'm happy to run a better set of tests to measure impact here, I just don't know what they are.

jasnell commented:
@mscdex ... with regards to performance, it's important just to note that this is largely just a formalization of a pattern we were already using. I wouldn't expect the performance to be much different than what we see currently, tho I do suspect the likelihood of bugs to go way down ;-)

trevnorris commented on src/aliased_buffer.h:
```diff
@@ -0,0 +1,193 @@
+    freeBuffer_(true) {
+    const v8::HandleScope handle_scope(this->isolate_);
+
+    const int sizeInBytes = sizeof(NativeT) * count;
```
`sizeof()` returns a `size_t`.

trevnorris commented on src/aliased_buffer.h:
```diff
@@ -0,0 +1,193 @@
+
+    // allocate native buffer
+    this->buffer_ = reinterpret_cast<NativeT*>(
+      calloc(count, sizeof(NativeT)));
```
Two problems here. 1) please use `node::UncheckedCalloc()` instead of using `calloc()` directly. 2) You're not checking if `this->buffer_ == nullptr` afterwards.

mike-kaufman replied:
Thanks @trevnorris, will do.  Is there an example of fail-fast behavior when out of memory occurs?  e.g., is there a specific error code to return?

jasnell replied:
Good catch. I tend to completely forget that we have those Calloc wrappers there.

trevnorris commented on src/aliased_buffer.h:
```diff
@@ -0,0 +1,193 @@
+    v8::Uint8Array>& backingBuffer) :
+    isolate_(isolate),
+    count_(count),
+    byte_offset_(byte_offset),
```
A `CHECK()` to verify that `byte_offset_` is aligned would be helpful.

mike-kaufman replied:
fixed.

addaleax replied:
Fwiw, that check isn't strictly necessary because the V8 TypedArray constructor checks it anyway

jasnell commented:
@trevnorris... see here for some of the background leading into this: https://github.com/nodejs/node-chakracore/issues/372

The short version is, the way the code is written now, it causes some headaches for the Chakra folks that this revised method is intended to ease. Given the problem, I think this is the best way to address it but I'd certainly love to have you review and comment on that! :-)

mike-kaufman commented:
@trevnorris  - 

> The patch doesn't follow general whitespace and line break rules that the remainder of core follows.

Linter is clean on osx for me.  Can you point out specifics?  Or is there a style guide I'm missing?

> Overall I'm -1 on this change. The abstraction adds more code than it removes, and it substantially adds more complexity. 

The key thing here is this simplifies cases where we need to observe writes to that program state.  e.g., take a look at https://github.com/nodejs/node-chakracore/pull/252/files.  The code here to "mark" these writes is scattered around in a way that is prone to breakage & merge conflicts.   The `AliasedBuffer` change lets us consolidate this at a single point in a maintainable/sane way. 

Happy to discuss other approaches here if you have ideas.

addaleax commented on src/aliased_buffer.h:
```diff
@@ -0,0 +1,194 @@
+    this->buffer_ = UncheckedCalloc<NativeT>(count);
+    if (this->buffer_ == nullptr) {
+      ABORT();
+    }
```
At this point you might as well use the checked kind of alloc? ;)

mike-kaufman replied:
fixed.

addaleax commented on src/aliased_buffer.h:
```diff
@@ -0,0 +1,194 @@
+
+    // allocate v8 ArrayBuffer
+    v8::Local<v8::ArrayBuffer> ab = v8::ArrayBuffer::New(
+      this->isolate_, this->buffer_, sizeInBytes);
```
tiny style nit: 4 spaces indentation for statement continuations (and we don't usually do `this->`)

mike-kaufman replied:
added 4 spaces for continuations & removed `this->` qualifiers.

addaleax commented on src/aliased_buffer.h:
```diff
@@ -0,0 +1,194 @@
+
+    // allocate v8 TypedArray
+    v8::Local<V8T> jsArray = V8T::New(ab, this->byte_offset_, count);
+    this->jsArray_ = v8::Global<V8T>(isolate, jsArray);
```
tiny style nit: snake_case (`js_array`, `js_array_`) for variables/member fields

mike-kaufman replied:
snake cased vars & memebers.  LMK if I missed anything.

addaleax commented on src/aliased_buffer.h:
```diff
@@ -0,0 +1,194 @@
+      return *this;
+    }
+
+    operator NativeT() {
```
This can be `const`, fwiw

mike-kaufman replied:
fixed

addaleax commented on src/aliased_buffer.h:
```diff
@@ -0,0 +1,194 @@
+  /**
+   *  Synonym for GetBuffer()
+   */
+  inline const NativeT* operator * () {
```
This can be `const` as well

mike-kaufman replied:
fixed

addaleax commented on src/env-inl.h:
```diff
@@ -346,7 +349,7 @@ inline Environment::~Environment() {
   delete[] heap_space_statistics_buffer_;
   delete[] http_parser_buffer_;
-  free(http2_state_buffer_);
+  delete  http2_state_;
```
*really* tiny style nit_ a single space after `delete`

mike-kaufman replied:
fixed

addaleax commented on src/env-inl.h:
```diff
@@ -457,7 +462,8 @@ inline double Environment::trigger_id() {
 inline double Environment::get_init_trigger_id() {
-  double* uid_fields = async_hooks()->uid_fields();
+  AliasedBuffer<double, v8::Float64Array> uid_fields =
+    async_hooks()->uid_fields();
```
You could use a reference (`AliasedBuffer<double, v8::Float64Array>&`) instead, right? I think that would allow you to get rid of `AliasedBuffer`'s copy constructor (and mark that as deleted), which seems like a good idea to me, I think

mike-kaufman replied:
yeah, you're right.  I'll fix...

mike-kaufman replied:
fixed

addaleax commented on src/node_http2.cc:
```diff
@@ -149,7 +93,8 @@ ssize_t Http2Session::OnCallbackPadding(size_t frameLen,
   if (object()->Has(context, env()->ongetpadding_string()).FromJust()) {
-    uint32_t* buffer = env()->http2_state_buffer()->padding_buffer;
+    AliasedBuffer<uint32_t, v8::Uint32Array> buffer =
+      env()->http2_state()->padding_buffer;
```
All of these could be references as well, right?

mike-kaufman replied:
fixed

addaleax commented on test/cctest/test_aliased_buffer.cc:
```diff
@@ -0,0 +1,220 @@
+  // read through the API
+  for (size_t i = 0; i < size; i++) {
+    NativeT v1 = (NativeT)aliasedBuffer[i];
+    NativeT v2 = (NativeT)aliasedBuffer.GetValue(i);
```
style nit: Can you use `static_cast` for these?

mike-kaufman replied:
removed cast, wasn't necessary.

addaleax commented on test/cctest/test_aliased_buffer.cc:
```diff
@@ -0,0 +1,220 @@
+    // NativeT actualValue = *(reinterpret_cast<NativeT*>(&v4->Value()));
+    // EXPECT_TRUE((NativeT)(v4->Value()) == oracle[i]);
+    // TODO(@mike-kaufman)  figure out how to compare the value in the v8 array
+    // Expect_TRUE(*v.ToLocal() == v8NumberCreateFunction(oracle[i]);
```
I am not sure what this code would mean, if I'm honest. Is this code for checking object identity between number values?

mike-kaufman replied:
I want to validate that the value read through JS APIs was the same value as what was written to the native array.  I just never found the right set of APIs and/or casts to make this happen.  Was hoping there was something straightforward that I was missing.

addaleax commented on test/cctest/test_aliased_buffer.cc:
```diff
@@ -0,0 +1,220 @@
+    EXPECT_TRUE(v2->IsNumber());
+    // v8::MaybeLocal<v8::Number> v3 = v2->ToNumber();
+    // v8::Local<v8::Number> v4 = v3.ToLocalChecked();
+    // NativeT actualValue = *(reinterpret_cast<NativeT*>(&v4->Value()));
```
I don't think this could work, you're taking a pointer to a stack-allocated return value and casting it around`.

`NativeT actualValue = static_cast<NativeT>(&v4->Value());`?

mike-kaufman replied:
I'll play around with that...

mike-kaufman replied:
OK, got this to work.  Think I was seeing seg fault earlier because I didn't pass `context` to `ToNumber`

addaleax commented:
> Overall I'm -1 on this change. The abstraction adds more code than it removes, and it substantially adds more complexity. This isn't a pattern that we need to use much, and I'm willing to bet that it's one we won't need to use much in the future.

@trevnorris This pattern solves a real problem, so if you have other solutions, feel free to voice them (but not doing so is kind of unproductive). Also, I would argue that the encapsulation that comes with this *does* remove complexity; it makes it more obvious to readers that a write to these ArrayBuffers is not just a memory write, but actually changes state for JS.

mike-kaufman commented:
Thanks @addaleax for the review!   I think I addressed all the feedback & rebased off of master.  LMK if I missed anything.  I didn't squash in case anyone wants to look at diffs, but I can do that.

jasnell commented:
@trevnorris ... I absolutely have to agree with @addaleax on this. This addresses a real problem even if we don't use this pattern extensively. Please reconsider your -1

addaleax commented:
@mike-kaufman Just occurred to me that a proper solution for the `delete http2_state;` problem might be to move the `Environment` constructor and destructor definition out of the header file and into `env.cc`, if you want to explore that. (You don’t need to, since you already figured out sth here)

jasnell commented:
ping @trevnorris

trevnorris commented on src/aliased_buffer.h:
```diff
@@ -0,0 +1,191 @@
+      free_buffer_(false) {
+    const v8::HandleScope handle_scope(isolate_);
+    // validate that the byte_offset is aligned with sizeof(NativeT)
+    CHECK_EQ(byte_offset & (sizeof(NativeT) - 1), 0);
```
This is missing a range check to make sure the byte length of this instance fits within the `backing_buffer`. e.g.:
```cc
CHECK_LE(sizeof(NativeT) * count, backing_buffer.GetArrayBuffer()->ByteLength() - byte_offset);
```

addaleax replied:
Fwiw, both that check and the existing check are also verified within V8

trevnorris replied:
@addaleax Possibly in a debug build, but I just created a module that creates a `Float64Array` and allowed me to have a count larger than the byte length of the array buffer.

mike-kaufman replied:
added check

trevnorris commented on src/aliased_buffer.h:
```diff
@@ -0,0 +1,191 @@
+  }
+
+  ~AliasedBuffer() {
+    if (free_buffer_ && buffer_ != NULL) {
```
`nullptr`

mike-kaufman replied:
fixed

trevnorris commented on src/aliased_buffer.h:
```diff
@@ -0,0 +1,191 @@
+  class Reference {
+   public:
+    Reference(AliasedBuffer<NativeT, V8T>* aliased_buffer, size_t index) :
+      aliased_buffer_(aliased_buffer),
```
please at least make the indentation style within the same file consistent. the majority of places in core (which I thought was part of the linter because we've been fairly consistent) would look like:
```cc
Reference(AliasedBuffer<NativeT, V8T>* aliased_buffer, size_t index)
  : aliased_buffer_(aliased_buffer),
    index_(index) {
```

trevnorris commented on src/env-inl.h:
```diff
@@ -445,7 +448,9 @@ inline std::vector<double>* Environment::destroy_ids_list() {
-  return ++async_hooks()->uid_fields()[AsyncHooks::kAsyncUidCntr];
+  async_hooks()->uid_fields()[AsyncHooks::kAsyncUidCntr] =
+    async_hooks()->uid_fields()[AsyncHooks::kAsyncUidCntr] + 1;
+  return async_hooks()->uid_fields()[AsyncHooks::kAsyncUidCntr];
```
is this only a style change, or does the previous operation not work anymore?

addaleax replied:
The previous operation doesn’t work anymore, we’d need to overload `operator++` for that, and from there it quickly starts to devolved into supporting all integer operations

jasnell commented:
How about we separate concerns here. Let's pull the async_hooks related changes out of this PR and ,move them into a separate PR. We can get this one landed then deal with the async_hooks specific changes separately.

trevnorris commented:
@mike-kaufman 
> The AliasedBuffer class provides a future placeholder where this
technique can be used, but writes can still be observed.

Can you elaborate on how this "future placeholder" will be used? I don't see how this PR does anything to help notification of state change when JS writes to memory.

addaleax commented on src/aliased_buffer.h:
```diff
@@ -0,0 +1,191 @@
+   *  Get value at position index
+   */
+  inline const NativeT GetValue(const size_t index) const {
+    CHECK_LT(index, count_);
```
@trevnorris I’m pretty sure the reported performance difference is due to this check, it’s the only visible difference in the generated output. I’d suggest wrapping it in `#if defined(DEBUG) && DEBUG`?

trevnorris replied:
confirmed. that is where the performance hit is coming from. wrapping it sounds good

mike-kaufman replied:
this and the SetValue check are now in debug-only builds.

addaleax commented on src/aliased_buffer.h:
```diff
@@ -0,0 +1,191 @@
+   *  Set position index to given value.
+   */
+  inline void SetValue(const size_t index, NativeT value) {
+    CHECK_LT(index, count_);
```
(same goes for this one, I assume)

mike-kaufman commented:
Sorry for the delayed response here, I've been off the grid for the last week or so.  I'll address latest feedback shortly. 

@trevnorris -

>Can you elaborate on how this "future placeholder" will be used? I don't see how this PR does anything to help notification of state change when JS writes to memory.

Short term, in the chakra fork, we'll add some logic in the `SetValue` method that will track anytime this state is modified.  Long term is TBD.  If need for something more general arises, you could have an API that allows callbacks to be registered that will get fired every time `SetValue` is called.  Irrespective of the details, with this change there's now a consolidated place where these writes occur, namely `SetValue`.

mike-kaufman commented on src/aliased_buffer.h:
```diff
@@ -0,0 +1,191 @@
+    const size_t sizeInBytes = sizeof(NativeT) * count;
+
+    // allocate native buffer
+    buffer_ = Calloc<NativeT>(count);
```
added `CHECK_GT(count, 0)`

mike-kaufman commented:
I think I addressed the latest feedback.  Please let me know if I missed anything.  

@jasnell - 

> How about we separate concerns here. Let's pull the AsyncWrap related changes out of this PR and ,move them into a separate PR.

I'm happy to do this, but I feel like we're close here?  LMK if not the case.

jasnell commented:
As long as @trevnorris and @addaleax are happy, I'm good.

trevnorris commented on src/aliased_buffer.h:
```diff
@@ -0,0 +1,199 @@
+    buffer_ = reinterpret_cast<NativeT*>(
+        const_cast<uint8_t*>(backing_buffer.GetNativeBuffer() + byte_offset));
+
+    v8::Local<v8::ArrayBuffer> ab = backing_buffer.GetArrayBuffer();
```
would it be worth it to assign `backing_buffer.GetNativeBuffer()` to a variable at the top, since it's used three times here? (not a blocker)

mike-kaufman replied:
fixed.  (assuming you meant `GetArrayBuffer()`).

trevnorris commented on src/aliased_buffer.h:
```diff
@@ -0,0 +1,199 @@
+  inline const NativeT GetValue(const size_t index) const {
+    #if defined(DEBUG) && DEBUG
+      CHECK_LT(index, count_);
+    #endif
```
nit (non-blocking): more common indentation is to have no whitespace before the `#if`.

mike-kaufman replied:
fixed

addaleax commented:
Fresh CI: https://ci.nodejs.org/job/node-test-commit/12329/

This should be ready.

jasnell commented:
Still LGTM

mike-kaufman commented:
Great!  Thanks @addaleax, @trevnorris, @jasnell!  Anything else I need to do here?  When can this get merged in?

jasnell commented:
If someone else doesn't beat me to it, I'll be merging it in later on this morning.

addaleax commented:
Landed in 7568579a6eb16af103846f988b29cc1af98039fd
